console-subscriber/src/visitors.rs

```diff
@@ -0,0 +1,369 @@
+        // don't care (yet?)
+    }
+
+    fn record_u64(&mut self, field: &tracing_core::Field, value: u64) {
+        if field.name() == Self::TASK_ID_FIELD_NAME {
+            self.id = Some(span::Id::from_u64(value));
+        }
+    }
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        if field.name() == "op" {
+            self.op = Some(match value {
+                Self::WAKE => WakeOp::Wake,
+                Self::WAKE_BY_REF => WakeOp::WakeByRef,
+                Self::CLONE => WakeOp::Clone,
+                Self::DROP => WakeOp::Drop,
+                _ => return,
+            });
+        }
+    }
+}
+
+impl PollOpVisitor {
+    pub(crate) const POLL_OP_EVENT_NAME: &'static str = "runtime.resource.poll_op";
+    const OP_NAME_FIELD_NAME: &'static str = "op_name";
+    const OP_READINESS_FIELD_NAME: &'static str = "readiness";
+    const OP_READINESS_READY: &'static str = "ready";
+    const OP_READINESS_PENDING: &'static str = "pending";
+
+    pub(crate) fn result(self) -> Option<(String, Readiness)> {
+        let op_name = self.op_name?;
+        let readiness = self.readiness?;
+        Some((op_name, readiness))
+    }
+}
+
+impl Visit for PollOpVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {}
+
+    fn record_str(&mut self, field: &tracing_core::Field, value: &str) {
+        match field.name() {
+            Self::OP_NAME_FIELD_NAME => {
+                self.op_name = Some(value.to_string());
+            }
+            Self::OP_READINESS_FIELD_NAME => {
+                self.readiness = Some(match value {
+                    Self::OP_READINESS_READY => Readiness::Ready,
+                    Self::OP_READINESS_PENDING => Readiness::Pending,
+                    _ => return,
+                });
+            }
+            _ => {}
+        }
+    }
+}
+
+impl StateUpdateVisitor {
+    pub(crate) const STATE_UPDATE_EVENT_NAME: &'static str = "runtime.resource.state_update";
+    const OP_STATE_FIELD_TYPE_ATTR_NAME: &'static str = "attribute_name";
+    const OP_STATE_FIELD_TYPE_VALUE: &'static str = "value";
+    const OP_STATE_FIELD_TYPE_UNIT: &'static str = "unit";
+    const OP_STATE_FIELD_TYPE_OP: &'static str = "op";
+    const UPDATE_OP_ADD: &'static str = "add";
+    const UPDATE_OP_SUB: &'static str = "sub";
+    const UPDATE_OP_OVR: &'static str = "ovr";
+
+    pub(crate) fn result(self, metadata_id: proto::MetaId) -> Option<AttributeUpdate> {
+        let name = self.name?;
+        let value = self.val?;
+        let op = self.op?;
+
+        let val = proto::Field {
+            metadata_id: Some(metadata_id),
+            name: Some(PbFieldName::StrName(name)),
+            value: Some(value),
+        };
+
+        Some(AttributeUpdate {
+            val,
+            op,
+            unit: self.unit,
+        })
+    }
+}
+
+impl Visit for StateUpdateVisitor {
+    fn record_debug(&mut self, _: &field::Field, _: &dyn std::fmt::Debug) {}
```

Comment:
should we also allow recording `Debug` fields for state updates? I can imagine a resource wanting to include some random bit of internal data as part of its tracked state. if we support `Debug` here we should probably assume that there's no `unit` and that the `op` is always "overwrite", the way we do for bool values...